reps/2023-03-10-batch-inference.md

```diff
@@ -0,0 +1,428 @@
+predictor = BatchPredictor.from_checkpoint(
+    checkpoint, MXNetPredictor, net=net
+)
+predictor.predict(dataset)
+```
+
+#### After
+```python
+class MXNetPredictor(Predictor):
+    def __init__(
+      self,
+      net: gluon.Block,
+    ):
+        self.net = net
+        super().__init__()
+
+    def _predict_numpy(
+        self,
+        data: Union[np.ndarray, Dict[str, np.ndarray]],
+        dtype: Optional[np.dtype] = None,
+    ) -> Dict[str, np.ndarray]:
+        # If `data` looks like `{"features": array([...])}`, unwrap the `dict` and pass
+        # the array directly to the model.
+        if isinstance(data, dict) and len(data) == 1:
+            data = next(iter(data.values()))
+
+        inputs = mx.nd.array(data, dtype=dtype)
+        outputs = self.net(inputs).asnumpy()
+
+        return {"predictions": outputs}
+
+dataset = ray.data.read_images(
+    "s3://anonymous@air-example-data-2/imagenet-sample-images", size=(224, 224)
+)
+
+def preprocess(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+    # (B, H, W, C) -> (B, C, H, W)
+    batch["image"] = batch["image"].transpose(0, 3, 1, 2)
+    return batch
+
+predictor = MXNetPredictor(gluon.model_zoo.vision.resnet50_v1(pretrained=True))
+dataset.map_batches(preprocess, batch_format="numpy").map_batches(predictor_creator, compute=ray.data.ActorPoolStrategy(size=8))
+```
+
+## Internal interfaces & implementation details
+1. Introduce a DeveloperAPI `BatchMappable` interface that `Preprocessor` and `Predictor` subclass. This interface will define the following:
+- Logic for the actual computation (either prediction or preprocessing)
+- Validating the user provided batch_format
+- Validating the user provided compute strategy
+
+```python
+@DeveloperAPI
+class BatchMappable(Callable[[T], U]):
+   """Subclass this if you want your class to be passable to Dataset.map_batches."""
+   
+    def __call__(self, batch: T) -> U:
+        """Transform the provided batch.
+
+        Args:
+            batch: The batch to transform.
+        """
+        
+        raise NotImplementedError
+
+    def setup(self):
+        """Additional setup logic to run.
+
+        If a callable class is being used, this function will be run in the `__init__`
+        of the Callable class.
+        """
+        pass
+       
+    def validate_batch_format(self, batch_format: Optional[str]) -> Optional[str]:
+        """Validate that the provided `batch_format` is compatible with this `BatchMappable`.
+
+        Raises an error if the provided `batch_format` is not supported with this `BatchMappable` 
+        (for example Pyarrow format for Predictors.)
+
+        If `batch_format` is "default", returns the preferred `batch_format` for this `BatchMappable`.
+
+        Args:
+            batch_format: The `batch_format` that is provided to `map_batches`.
+
+        Returns:
+            The batch format to use.
+        """
+       return None
+
+    def validate_compute_strategy(self, strategy):
```

Comment:
Shall we have a generic validate_args(kwargs) function instead of adding them one at a time?

Reply:
I think it's fine for now since it's just 2 (batch format, compute strategy). If we add more, then we can look into consolidating in the future.